webapp/src/extensionsBrowser.tsx

```diff
@@ -0,0 +1,478 @@
+import * as React from "react";
+import * as core from "./core";
+import * as workspace from "./workspace";
+import * as pkg from "./package";
+
+import { MenuBar } from "../../react-common/components/controls/MenuBar";
+import { Button } from "../../react-common/components/controls/Button";
+import { workerOpAsync } from "./compiler";
+import { SearchInput } from "./components/searchInput";
+import { useState, useEffect } from "react";
+import { ImportModal } from "../../react-common/components/extensions/ImportModal";
+import { DeleteConfirmationModal } from "../../react-common/components/extensions/DeleteConfirmationModal";
+import { ExtensionCard } from "../../react-common/components/extensions/ExtensionCard";
+
+type ExtensionMeta = pxtc.service.ExtensionMeta;
+const emptyCard = { name: "", loading: true }
+
+interface ExtensionsProps {
+    isVisible: boolean;
+    hideExtensions: () => void;
+    header: pxt.workspace.Header;
+    reloadHeaderAsync: () => Promise<void>;
+}
+
+enum TabState {
+    Recommended,
+    Installed,
+    InDevelopment
+}
+
+export const ExtensionsBrowser = (props: ExtensionsProps) => {
+
+    const [searchFor, setSearchFor] = useState("");
+    const [allExtensions, setAllExtensions] = useState(fetchBundled());
+    const [extensionsToShow, setExtensionsToShow] = useState([]);
+    const [selectedTag, setSelectedTag] = useState("");
+    const [currentTab, setCurrentTab] = useState(TabState.Recommended);
+    const [showImportExtensionDialog, setShowImportExtensionDialog] = useState(false);
+    const [installedExtensions, setInstalledExtensions] = useState([])
+    const [lastVisibleState, setLastVisibleState] = useState(props.isVisible)
+    const [deletionCandidate, setDeletionCandidate] = useState(undefined)
+    const [preferredExts, setPreferredExts] = useState([])
+    const [extensionTags, setExtensionTags] = useState(new Map<string, pxt.RepoData[]>())
+
+    if (lastVisibleState != props.isVisible) {
+        updateInstalledExts();
+        updateExtensionTags();
+        setLastVisibleState(props.isVisible)
+        updatePreferredExts();
+        if (!props.isVisible) {
+            setCurrentTab(TabState.Recommended)
+            setSearchFor("")
+            setSelectedTag("")
+        }
+    }
+
+
+    useEffect(() => {
+        updateInstalledExts();
+    }, [])
+
+    useEffect(() => {
+        if (searchFor && searchFor != "") {
+            if (searchFor.indexOf("/") >= 0) {
+                searchForGithubAsync();
+            } else {
+                workerOpAsync("extensionSearch", {
+                    search: {
+                        term: searchFor
+                    },
+                    extensions: {
+                        srcs: Array.from(allExtensions.values())
+                    }
+                }).then(e => {
+                    setExtensionsToShow(e)
+                    setSelectedTag("")
+                })
+            }
+        }
+    }, [searchFor])
```

Comment:
Does `allExtensions` need to be listed in dependencies?

Reply:
I think no, adding it causes an infinite loop